### PR TITLE
unittests and fixes for gtime2str

### PR DESCRIPTION
**Pull request by icing**
x506asn1.c: fix issues in GTime2str() and add unit test cases to verify correct behaviour.

### CODE REVIEW

File: tests/unit/unit1656.c
Comment (dubek) [inline tests/unit/unit1656.c:86]: Consider adding a test were the string doesn't end with `\0` (that is, `end` points to a digit or a dot or something like that).

File: lib/vtls/x509asn1.c
Comment (dubek) [inline lib/vtls/x509asn1.c:563]:
```suggestion
/* used by unit1656.c */
```

File: tests/unit/unit1656.c
Comment (bagder) [reply tests/unit/unit1656.c:86]: Right, add a length-field in the struct for each case or something so that you can make a date string end in the middle and it should stop parsing there.

File: lib/vtls/x509asn1.c
Comment (icing) [reply lib/vtls/x509asn1.c:563]: Right, needed to use another number and now there are relics. Thanks for catching this.

File: tests/unit/unit1656.c
Comment (icing) [reply tests/unit/unit1656.c:86]: Which begs the question: what would be the expected test result on this?

Comment (icing) [reply tests/unit/unit1656.c:86]: @bagder and it should fail any 0 between being and end?

Comment (bagder) [reply tests/unit/unit1656.c:86]: Well, if the string until the end if a valid string then that is fine and that's the date, even if it continues looking like a date beyond the end. If it is not a valid date before the end, then it's a failure.

**Review by bagder — APPROVED** [review]
Are you fine with me merging this as-is so we can have it in the 8.9.1 release?

**Comment by icing** [comment]
> Are you fine with me merging this as-is so we can have it in the 8.9.1 release?

Yes, please. Let's improve this afterwards.

File: lib/vtls/x509asn1.c
Comment (bagder) [reply lib/vtls/x509asn1.c:563]: maybe commit this correction at least before merge?

**Comment by bagder** [comment]
I just committed the comment fix, I intend to merge this once all/most CI jobs are green.

**Comment by dubek** [comment]
If you think it's good idea, I think I can work on a PR which adds fuzz entrypoints for GTime2str and Curl_extract_certinfo in curl-fuzzer project (not this week, though). It's a bit tricky because it looks like we build curl with openssl over there (which means no Curl_extract_certinfo ...). But if some distros are changing to build curl with gnutls, maybe we should have some fuzzing power spent over there.

**Comment by bagder** [comment]
@dubek it is a really good idea, and as you say that requires a build that uses one of the TLS backends for which we use the x509asn1 parser code. But I see no downside in doing that.

**Comment by bagder** [comment]
I will merge this PR now then so that we can have this fix in the release tomorrow. We can still polish things further in new PRs to merge later.

**Pull request closed without merging**
